### ISSUE TITLE
CellCycleScore function (AddModuleScore) function in Seurat v3

### ISSUE THREAD

**Issue by chlee-tabin** [satijalab/seurat#1227]
This is related to the closed issue:

https://github.com/satijalab/seurat/issues/1181#issue-415260625

I essentially encounter the same issue and I think it is due to the different behavior of the `Hmisc::cut2` function (in Seurat v2) and `ggplot2::cut_number` used in Seurat v3.

Specifically, the issue arises when there are too many genes that essentially have no counts in the entire dataset. Where Seurat v2 `cut2` function just does not generate error and creates a bin, the `cut_number` generates the insufficient data values to produce 24 bins error.

I haven't look very carefully what would be the best way to resolve it, but as it stands, I have to trim features out from the seurat object to run this function.

And as a side note, the `pbmc_small` appears to have no genes related to the `cc.genes` data (too small) such that the example for `CellCycleScore` breaks.

**Comment by Sophia409** [comment]
@chlee-tabin Hello, I also encounterred this Seurat v3.0 CellCycleScoring Error.Do you have any solutions now?
```

> cc.genes <- readLines(con = "D:/personal data/Yuhong Zhang/下丘脑课题/cell_cycle_vignette_files/regev_lab_cell_cycle_genes.txt")`
> cc.genes <- str_to_title(cc.genes)
> s.genes <- cc.genes[1:43]
> g2m.genes <- cc.genes[44:97]
> length(x = which(x = s.genes %in% rownames(x = GetAssayData(tc11))))
[1] 42
> length(x = which(x = g2m.genes %in% rownames(x = GetAssayData(tc11))))
[1] 52
> tc11 <- CellCycleScoring(object = tc11, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
Error: Insufficient data values to produce 24 bins.
```

**Comment by Thessentials** [comment]
I agree with @chlee-tabin. I also ran mine with ```min.cells=1``` in order to remove genes with no expression prior to normalizing. I think this works because you will be unlikely to get duplicate values that cause the ggplot2::cut_number to struggle creating the correct number of bins.

**Comment by satijalab** [comment]
Thanks! We should make sure that the function works when running pbmc_small, and perhaps introduce error-checking for the case where genes are expressed in no cells

**Comment by eisascience** [comment]
I got in to the code here.... the problem is with the code in 
AddModuleScore which uses 
  assay <- assay %||% assay.old

What and where does  %||% come from?

**Comment by eisascience** [comment]
the bin 24 comes from , nbin = 24 in the AddModuleScore but changing it still doesn't do anything because of I think of %||%

**Comment by Sophia409** [comment]
@chlee-tabin @Thessentials Thank you,  I ran my data with min.cells=1 and it worked.

**Comment by eisascience** [comment]
The min.cells=1 has not worked for me; I came up with a hacky work-around.

In the CellCycleScoring .....add this to replace the AddModueScore only 

```{r }
   StartNBin = 24
  object.cc = NA

  while(class(object.cc)=="try-error" || is.na(object.cc[1])){

    object.cc <- try(AddModuleScore(object = seuratObj, features = features, nbin = StartNBin,
                                    name = name,
                                    ctrl = min(vapply(X = features, FUN = length,
                                                      FUN.VALUE = numeric(length = 1)))))
    StartNBin = round(StartNBin/2)

  }; print(StartNBin)

```

**Comment by chlee-tabin** [comment]
@eisascience Based on your code (which as far as I understand halves the bins progressively to reach an non-error) wondering what your eventual bin size was that worked. Is your dataset UMI-based counts? 

Suspecting that your dataset is so sparse with only very few counts for all genes or the feature counts were few, so that it was breaking. As far as I understand, the logic behind the cell cycle scoring is to identify random gene expressions that match the range of expressions for the cell cycle genes (the original paper with 25 bins) so reducing the bin size would make the values of scoring less robust.

**Comment by eisascience** [comment]
Thanks for the feedback. It was 3 that worked. Now that I have ran the code and it returned 3, I agree that I did not like my own solution. That said, my pipeline worked with Seurat 2, and I recently switched to 3 which caused several pipeline breaks.  And yes it is UMI-based. However you may be correct about the sparsity. I will have to investigate with additional datasets.

**Comment by andrewwbutler** [comment]
I've push a fix for this as of cbf8759 that should fix the binning issue. Let us know if that resolves the issues here. 

Re the `pbmc_small` example - it's clearly stated in the documentation that the example is not meant to actually be run on that dataset and links to a more appropriate one. We included the example code mostly to make CRAN happy as they want examples for all/most functions. 

```
## Not run: 
# pbmc_small doesn't have any cell-cycle genes
# To run CellCycleScoring, please use a dataset with cell-cycle genes
# An example is available at http://satijalab.org/seurat/cell_cycle_vignette.html
pbmc_small <- CellCycleScoring(
  object = pbmc_small,
  g2m.features = cc.genes$g2m.genes,
  s.features = cc.genes$s.genes
)
head(x = pbmc_small@meta.data)

## End(Not run)
```

**Comment by crebli** [comment]
> The min.cells=1 has not worked for me; I came up with a hacky work-around.
> 
> In the CellCycleScoring .....add this to replace the AddModueScore only
> 
> ```
>    StartNBin = 24
>   object.cc = NA
> 
>   while(class(object.cc)=="try-error" || is.na(object.cc[1])){
> 
>     object.cc <- try(AddModuleScore(object = seuratObj, features = features, nbin = StartNBin,
>                                     name = name,
>                                     ctrl = min(vapply(X = features, FUN = length,
>                                                       FUN.VALUE = numeric(length = 1)))))
>     StartNBin = round(StartNBin/2)
> 
>   }; print(StartNBin)
> ```

Error : Insufficient data values to produce 24 bins.
Error : Insufficient data values to produce 12 bins.
[1] 3
Warning message:
In is.na(object.cc[1]) : is.na()不适用于类别为'S4'的非串列或非矢量
